### ISSUE TITLE
Possible error?

### ISSUE THREAD

**Issue by alif898** [j0sephsasson/numpy-NN#1]
https://github.com/j0sephsasson/numpy-NN/blob/0e8e2ce49269c4ea2c9f844a3f606322f69c9fc9/model/model.py#L90

Hi, should this be something like this instead?

`self.memory[i] = {'inputs':A_prev, 'Z':Z_curr}`

Currently, 
https://github.com/j0sephsasson/numpy-NN/blob/0e8e2ce49269c4ea2c9f844a3f606322f69c9fc9/model/model.py#L107
https://github.com/j0sephsasson/numpy-NN/blob/0e8e2ce49269c4ea2c9f844a3f606322f69c9fc9/model/model.py#L108

These two lines will only access the same values within self.memory again and again, as new values are appended instead of replacing the old ones.

**Comment by j0sephsasson** [comment]
Hello, thank you for reaching out! 

Answer 1: self.memory.append({'inputs':A_prev, 'Z':Z_curr}) --> this is valid because memory is a list, we need to call the 'append' method to add the new values.

Answer 2: when performing backprop, we want to access the same input and activation elements created during the forward pass.

**Comment by rballachay** [comment]
Hi @j0sephsasson, first off, thank you for the code! I read your article and found it very helpful. However, I am pretty sure this is still a bug. Every loop of the network, the memory and gradient get longer, however you only ever access the first few elements of it:

```python
  def _forwardprop(self, data):
      A_curr = data
      
      for i in range(len(self.params)):
            A_prev = A_curr
            A_curr, Z_curr = self.network[i].forward(inputs=A_prev, weights=self.params[i]['W'], 
                                           bias=self.params[i]['b'], activation=self.architecture[i]['activation'])
            
            self.memory.append({'inputs':A_prev, 'Z':Z_curr})

  def _update(self, lr=0.01):
      for idx, layer in enumerate(self.network):
            self.params[idx]['W'] -= lr * list(reversed(self.gradients))[idx]['dW'].T  
            self.params[idx]['b'] -= lr * list(reversed(self.gradients))[idx]['db']
```

This means that the network only converges but rarely, if the starting gradient and learning rate are set just right. You can test this by initializing as I have proposed here:

```python
    def _forwardprop(self, data):
        self.memory = []  <----- NEED TO RE-INITIALIZE HERE

        A_curr = data
        
        for i in range(len(self.params)):
            A_prev = A_curr
            A_curr, Z_curr = self.network[i].forward(inputs=A_prev, weights=self.params[i]['W'], 
                                           bias=self.params[i]['b'], activation=self.architecture[i]['activation'])
            
            self.memory.append({'inputs':A_prev, 'Z':Z_curr})
            
        return A_curr

    def _backprop(self, predicted, actual):
        self.gradients = []   <----- NEED TO RE-INITIALIZE HERE

        num_samples = len(actual)
        
        ## compute the gradient on predictions
        dscores = predicted
        dscores[range(num_samples),actual] -= 1
        dscores /= num_samples
        
        dA_prev = dscores
        
        for idx, layer in reversed(list(enumerate(self.network))):
            dA_curr = dA_prev
            
            A_prev = self.memory[idx]['inputs']
            Z_curr = self.memory[idx]['Z']
            W_curr = self.params[idx]['W']
            
            activation = self.architecture[idx]['activation']

            dA_prev, dW_curr, db_curr = layer.backward(dA_curr, W_curr, Z_curr, A_prev, activation)

            self.gradients.append({'dW':dW_curr, 'db':db_curr})

```
You can confirm that this converges 100% of the time + much faster than the prior version

**Comment by alif898** [comment]
Hi @rballachay thanks for sharing your solution, yes I think this was the same issue I faced as well, where we kept accessing the first few elements of memory throughout.

**Comment by j0sephsasson** [comment]
@rballachay, great find. You are absolutely correct! Feel free to make a PR :)
